File: backend/app/api/ferry.py
"""待引渡 HTTP 路由（M5）。"""

from __future__ import annotations

from fastapi import APIRouter, Depends

from app.core.deps import get_current_user, get_ferry_service
from app.db.models import User
from app.schemas.common import success
from app.schemas.mentor import FerryRescueRequest
from app.services.ferry_service import FerryService

router = APIRouter(prefix="/ferry", tags=["ferry"])


@router.get("/me", response_model=None)
async def ferry_me(
    svc: FerryService = Depends(get_ferry_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """待引渡状态与倒计时。"""
    return success(await svc.get_me(current_user))


@router.get("/rescue-targets", response_model=None)
async def ferry_rescue_targets(
    category: str = "universal",
    svc: FerryService = Depends(get_ferry_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """
    引渡救援名单。

    category: universal=普渡众生（道友）· sect=同门 · kin=亲友（道友/道侣/师徒/炉鼎）
    """
    return success(await svc.list_rescue_targets(current_user, category=category))


@router.post("/self-rescue", response_model=None)
async def ferry_self_rescue(
    svc: FerryService = Depends(get_ferry_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """自救。"""
    return success(await svc.self_rescue(current_user))


@router.post("/rescue", response_model=None)
async def ferry_social_rescue(
    body: FerryRescueRequest,
    svc: FerryService = Depends(get_ferry_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """普渡/道友 / 同门 / 亲友引渡。"""
    return success(
        await svc.social_rescue(
            current_user,
            target_character_id=body.target_character_id,
            target_name=body.target_name,
            mode=body.mode,
        ),
    )


@router.post("/enter-reincarnation", response_model=None)
async def ferry_enter_reincarnation(
    svc: FerryService = Depends(get_ferry_service),
    current_user: User = Depends(get_current_user),
) -> dict:
    """自选轮回。"""
    return success(await svc.enter_reincarnation(current_user))

File: frontend/src/api/ferry.ts
/**
 * M5 / M7 L6 待引渡 API：自救 / 社交引渡 / 自选进入轮回。
 */
import { http } from './http'
import { envelopeFromAxiosError } from './envelope'
import type { ApiResponse } from '../types/api'
import type {
  FerryMutationResult,
  FerryPublic,
  FerryRescueCategory,
  FerryRescueRequest,
  FerryRescueTargetsPayload,
} from '../types/ferry'

/** GET /ferry/me */
export async function fetchFerry(): Promise<ApiResponse<FerryPublic | null>> {
  try {
    const response = await http.get<ApiResponse<FerryPublic | null>>('/ferry/me')
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<FerryPublic | null>(error)
  }
}

/**
 * GET /ferry/rescue-targets — 按类别列出待引渡目标。
 *
 * @param category - universal | sect | kin
 */
export async function fetchRescueTargets(
  category: FerryRescueCategory = 'universal',
): Promise<ApiResponse<FerryRescueTargetsPayload>> {
  try {
    const response = await http.get<ApiResponse<FerryRescueTargetsPayload>>(
      '/ferry/rescue-targets',
      { params: { category } },
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError(error)
  }
}

/** POST /ferry/self-rescue */
export async function selfRescue(): Promise<ApiResponse<FerryMutationResult>> {
  try {
    const response = await http.post<ApiResponse<FerryMutationResult>>(
      '/ferry/self-rescue',
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<FerryMutationResult>(error)
  }
}

/**
 * POST /ferry/rescue — 普渡/同门/亲友引渡（救援者支付）。
 *
 * @param body - mode + 目标道号或角色 id
 */
export async function socialRescue(
  body: FerryRescueRequest,
): Promise<ApiResponse<FerryMutationResult>> {
  try {
    const response = await http.post<ApiResponse<FerryMutationResult>>(
      '/ferry/rescue',
      body,
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<FerryMutationResult>(error)
  }
}

/** POST /ferry/enter-reincarnation */
export async function enterReincarnation(body?: {
  confirm?: boolean
}): Promise<ApiResponse<FerryMutationResult>> {
  try {
    const response = await http.post<ApiResponse<FerryMutationResult>>(
      '/ferry/enter-reincarnation',
      { confirm: true, ...body },
    )
    return response.data
  } catch (error: unknown) {
    return envelopeFromAxiosError<FerryMutationResult>(error)
  }
}
